[google_maps_flutter] custom marker size

jokerttu opened the pull request:
This PR adds and improves marker scaling support for Android, iOS and Web.
With this change, custom marker icons are drawn with same logical size on all platforms, keeping the visual style as close as possible.

This PR is created from previous PR: https://github.com/flutter/plugins/pull/6805

To avoid breaking change, this PR makes the following changes while keeping the old behavior intact:
- Deprecates BitmapDescriptor.fromAssetImage in favor of BitmapDescriptor.createFromAsset
- Deprecates BitmapDescriptor.fromBytes in favor of BitmapDescriptor.createFromBytes

**Android:**
![image](https://github.com/flutter/packages/assets/5219613/c5362261-a487-4917-967a-5d2bf5a38797)

**Web:**
![image](https://github.com/flutter/packages/assets/5219613/4a66219b-1ee0-40ac-ae84-2cf741a64cad)

Resolves https://github.com/flutter/flutter/issues/34657

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

ilDeffo commented:
Any update on when this will be merged?

jokerttu commented on packages/google_maps_flutter/google_maps_flutter/example/integration_test/google_maps_test.dart:
```diff
@@ -944,17 +944,21 @@ void main() {
     const double pixelRatio = 2;
     const ImageConfiguration imageConfiguration =
         ImageConfiguration(devicePixelRatio: pixelRatio);
-    final BitmapDescriptor mip = await BitmapDescriptor.fromAssetImage(
```
These tests are removed from this plugin as this functionality is part of the platform_interface.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_web/pubspec.yaml:
```diff
@@ -23,6 +23,7 @@ dependencies:
     sdk: flutter
   google_maps: ^6.1.0
   google_maps_flutter_platform_interface: ^2.2.2
+  image: ^4.0.13
```
Yes, we can remove this library and functionality to have auto-scaled images for web.
If the size is not given, marker size will differ by default between web and other platforms.
As reading the image size from the asset is not optimal on the web platform, removing this feature is actually good solution as developers can do it by themselves on need.
I'll update the documentation as well to inform the need for providing size.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -219,6 +254,35 @@ - (UIImage *)scaleImage:(UIImage *)image by:(id)scaleParam {
+                        scale:(image.scale * (size.width / (image.size.width * image.scale)))];
+    } else {
+      UIGraphicsBeginImageContextWithOptions(size, NO, 1.0);
+      [image drawInRect:CGRectMake(0, 0, size.width, size.height)];
```
Yes, this is intended behavior.
I'll update the docs that size need to be given in proper aspect ration.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_web/README.md:
```diff
@@ -50,3 +52,25 @@ Indoor and building layers are still not available on the web. Traffic is.
+
+## Custom marker icons
+
+In order to achieve the best possible performance when using custom marker images on the web platform, the `size` parameter should be used to scale the marker when using `BitmapDescriptor.createFromAsset` or `BitmapDescriptor.createFromBytes` methods.
```
Auto scaling for web was removed and this part of the README.md is updated

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -219,6 +254,35 @@ - (UIImage *)scaleImage:(UIImage *)image by:(id)scaleParam {
   return image;
 }
 
+- (UIImage *)scaleImage:(UIImage *)image scale:(CGFloat)scale {
```
scaleImage:by: does scaling wrong and was used with old now deprecating fromAsset and fromBytes versions of the BitmapDescriptor. The functionality differs in two places, how scale value is checked, and how the scale is set to the image.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -85,6 +98,81 @@ private static BitmapDescriptor getBitmapFromBytes(List<?> data) {
+
+    return BitmapDescriptorFactory.fromAsset(asset);
+  }
+
```
I used quite a much time to figure out how to build unit tests for these two methods `getBitmapFromBytes` and `getBitmapFromAsset`, and didn't found a good way to do this as there is few issues:
`BitmapDescriptorFactory` seems to be not available for Unit tests under `RobolectricTestRunner`.
Also mocking the FlutterLoader seems to be impossible it is already initialized for the FlutterInjector, so mocking the asset loading seems quite complicated issue to fix.
I need to put this for hold for now; and I might be able to proceed with this issue after July. Any ideas to mock the static methods of BitmapDescriptorFactory or how to pass assets to the test apk are welcome in meantime.

stuartmorgan replied:
> `BitmapDescriptorFactory` seems to be not available for Unit tests under `RobolectricTestRunner`.

Generally we solve issues of classes not being available in unit tests by wrapping them, and having a visible-for-testing constructor that lets us use DI to provide a test implementation of that wrapper, while the normal constructor calls that constructor with a default implementation using the class being wrapped.

> Also mocking the `FlutterLoader` seems to be impossible it is already initialized for the `FlutterInjector`

Similarly, you can DI with a minimal wrapper around asset loading.

cyanglaz commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -160,52 +160,89 @@ - (UIImage *)extractIconFromData:(NSArray *)iconData
 }
 
+// Used by deprecated fromBytes functionality.
+// Can be removed when deprecated image bitmap types are removed from platform interface.
```
Can you link to the exact type signature here? It is hard to find out what `deprecated image bitmap types` are when we want to clean up the code later. It would be nice to create a Github issue to track this and add a TODO here.

jokerttu replied:
Update the docs for this method to describe the deprecation better

cyanglaz commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -219,6 +256,66 @@ - (UIImage *)scaleImage:(UIImage *)image by:(id)scaleParam {
 
+/**
+ * Creates a scaled version of the provided UIImage based on a specified scale factor. If the scale
+ * factor is significantly different from the image's current scale, a new UIImage object is
```
What does "significantly" mean in this context?
Maybe rephrase a little saying if the scale factor is almost identical (a tiny epsilon-delta is ignored) to the current scale, return the original image.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -219,6 +256,66 @@ - (UIImage *)scaleImage:(UIImage *)image by:(id)scaleParam {
+- (CGSize)scaleSizeToInt:(CGSize)size withFactor:(CGFloat)scale {
+  return CGSizeMake((int)(size.width * scale), (int)(size.height * scale));
+}
+
```
It looks like I forgot to flag it in the previous round, but the changes in this file also need native unit tests.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/ios/RunnerTests/ExtractIconFromDataTests.m:
```diff
@@ -0,0 +1,175 @@
+  XCTAssertNotNil(resultImage);
+}
+
+- (void)testExtractIconFromDataBytesNoSacaling {
```
Typo: Scaling

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios13/ios/RunnerTests/ExtractIconFromDataTests.m:
We don't need a second copy of this test file; the tests in `ios13` are for iOS-13+-specific codepaths.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/bitmap.dart:
```diff
@@ -4,12 +4,27 @@
+  /// to keep marker sizes same between platforms and devices.
+  auto,
+
+  /// Render marker to the map as without scaling, this can be used if the image
```
Nit: Replace the `,` after scaling with a `.` and two newlines (and then capitalize `This`).

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_web/README.md:
```diff
@@ -72,3 +74,25 @@ Indoor and building layers are still not available on the web. Traffic is.
+
+### Custom marker icons
+
+Images are not automatically scaled according to the imageAspectRatio value of the BitmapDescriptor in web, so the `size` parameter must be used to scale the marker when using the `BitmapDescriptor.createFromAsset` or `BitmapDescriptor.createFromBytes` methods.
```
Nit: backticks around `imageAspectRatio` and `BitmapDescriptor`.

stuartmorgan reviewed (commented):
A few minor nits, on the cross-platform parts, but generally they look good, and the native test coverage looks good now from a high level, so this should be ready for full platform reviews by @cyanglaz, @ditman, and @reidbaker.

stuartmorgan commented:
@cyanglaz, @ditman, @reidbaker: ping on the platform portions of this review.

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -34,10 +37,8 @@
-  //  when it's available in Stable channel: https://github.com/flutter/flutter/issues/70923.
-  @SuppressWarnings("deprecation")
-  private static BitmapDescriptor toBitmapDescriptor(Object o) {
+  private static BitmapDescriptor toBitmapDescriptor(
```
What do you think about doing the data.get(#) operations before the business logic and putting them into an object with more type safety for the rest of this code? There are several places where a named variable would make the code easier to ready over `toString(data.get(1)` for example

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -85,6 +103,119 @@ private static BitmapDescriptor getBitmapFromBytes(List<?> data) {
+    }
+  }
+
+  /**
```
This is a great javadoc.

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -698,4 +844,52 @@ static Tile interpretTile(Map<String, ?> data) {
+    }
+  }
+
+  static class FlutterInjectorWrapper {
```
Should this also be marked as @visibleForTesting?

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -698,4 +844,52 @@ static Tile interpretTile(Map<String, ?> data) {
+     * @return the lookup key for the asset.
+     */
+    @VisibleForTesting
+    public String getLookupKeyForAsset(String assetName) {
```
Can you add a @nullable or @nonNull annotation to the assetName?

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/ConvertTest.java:
```diff
@@ -26,4 +72,163 @@ public void ConvertToPointsConvertsThePointsWithFullPrecision() {
+
+  @Test
+  public void GetBitmapFromBytesAutoAndSize() throws Exception {
+    String base64Image =
```
Can you add a comment for how you generated this string and how others can create a new version if they need to later?

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/ConvertTest.java:
```diff
@@ -26,4 +72,163 @@ public void ConvertToPointsConvertsThePointsWithFullPrecision() {
+  @Test
+  public void GetBitmapFromBytesNoScaling() throws Exception {
+    String base64Image =
+        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg==";
```
I think this is the same string that is used in the previous test. If so can you make it a constant defined at the top of the class?

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/ConvertTest.java:
```diff
@@ -4,13 +4,59 @@
+import org.robolectric.annotation.Config;
 
+@RunWith(RobolectricTestRunner.class)
+@Config(sdk = Build.VERSION_CODES.P)
```
For my own understanding can you tell me why we need to run this on only api level 28? My gut is that some tests should be run on different api levels.

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/MarkersControllerTest.java:
```diff
@@ -21,16 +24,32 @@
+import org.robolectric.annotation.Config;
 
+@RunWith(RobolectricTestRunner.class)
+@Config(sdk = Build.VERSION_CODES.P)
```
SImilar question here can you let me know why this should only run on api level 28?

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:
```diff
@@ -243,16 +243,17 @@ gmaps.Size? _gmSizeFromIconConfig(List<Object?> iconConfig, int sizeIndex) {
 
 // Converts a [BitmapDescriptor] into a [gmaps.Icon] that can be used in Markers.
-gmaps.Icon? _gmIconFromBitmapDescriptor(BitmapDescriptor bitmapDescriptor) {
+Future<gmaps.Icon?> _gmIconFromBitmapDescriptor(
+    BitmapDescriptor bitmapDescriptor) async {
```
Does this method need to be `async`? I don't see any `await`s in it? This should allow this change to simplify a bunch of async/awaits in the rest of the code, I think?

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:
```diff
@@ -283,17 +284,51 @@ gmaps.Icon? _gmIconFromBitmapDescriptor(BitmapDescriptor bitmapDescriptor) {
         ..size = size
         ..scaledSize = size;
     }
+  } else if (iconConfig[0] == 'asset') {
```
The implementations of the new `asset` and `bytes` are very similar to the existing ones. Any possibility to unify them?

ditman commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/bitmap.dart:
```diff
@@ -175,6 +230,116 @@ class BitmapDescriptor {
+    }
+  }
+
   final Object _json;
```
These _json-backed objects are ugly, and force us to write ugly code in the `convert.dart`.

Is there any way that this could be separated into something like:

`abstract class MapBitmap extends BitmapDescriptor` (so it can be passed around) and then have:

`class AssetMapBitmap extends MapBitmap` and `class BytesMapBitmap extends MapBitmap` which can be created as usual?

Then the changes in the convert.dart and elsewhere would be minimal; just a type check and a cast to use the right subtype and its accessors?

It seems the only public method on all of these are `.toJson`, which can probably be skipped for the new classes.

(Not sure how complex this is for other platforms, but I suspect that the "json" object is "convenient" for the method channel serialization?)

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/README.md:
```diff
@@ -1,5 +1,7 @@
 # google_maps_flutter_web
 
+<?code-excerpt path-base="excerpts/packages/google_maps_flutter_web_integration_tests"?>
```
There should be a code block underneath this code-excerpt directive? Maybe code excerpts haven't been generated for this package?

ditman reviewed (commented):
A few comments about the web, and a small rant about the _json-backed BitmapDescriptor in Dart (which I'd love to see gone if possible :))

gmackall commented:
@jokerttu It looks like there are still some outstanding comments on this PR, is this something you are still actively working on?

jokerttu commented:
> @jokerttu It looks like there are still some outstanding comments on this PR, is this something you are still actively working on?

Yes, I am actively working on addressing the outstanding feedback. I will make the necessary updates to the PR in the near future.

camsim99 commented:
I am marking the PR as a Draft to remove it from our queue. When you're actively working on this again, please feel free to mark the PR as ready for review.

goderbauer commented:
(triage): @jokerttu Do you still have plans to come back to this PR in the near future?

jokerttu commented:
> (triage): @jokerttu Do you still have plans to come back to this PR in the near future?

I am currently rewriting updated version following the request from @ditman https://github.com/flutter/packages/pull/4055/files#r1326628458
Revisited version coming soon.

jokerttu reviewed (commented):
Refactored the functionality to separate classes as requested.
Instead of `BitmapDescriptor.createFromAsset` there is `AssetMapBitmap` and `AssetMapBitmap.fromMipmaps`.
and Instead of `BitmapDescriptor.createFromBytes` there is `BytesMapBitmap`

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -219,6 +261,67 @@ - (UIImage *)scaleImage:(UIImage *)image by:(id)scaleParam {
+ * @param scale The scale factor to apply to the width and height of the CGSize.
+ * @return CGSize Returns the scaled CGSize with width and height as integers.
+ */
+- (CGSize)scaleSizeToInt:(CGSize)size withFactor:(CGFloat)scale {
```
nit: `scaleSizeAndRoundToInt`

Btw why do we want int here?

jokerttu replied:
This conversion is used for calculating image bitmap width and height.

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -219,6 +261,67 @@ - (UIImage *)scaleImage:(UIImage *)image by:(id)scaleParam {
+      return [self scaleImage:image
+                    withScale:(image.scale * (size.width / (image.size.width * image.scale)))];
+    } else {
+      UIGraphicsBeginImageContextWithOptions(size, NO, 1.0);
```
can you add comment on each parameter names?

jokerttu replied:
replaced Deprecated UIGraphicsBeginImageContextWithOptions with UIGraphicsImageRenderer (available since ios10).

jokerttu replied:
Update the functionality to use UIGraphicsImageRendererFormat instead as UIGraphicsBeginImageContextWithOptions was deprecated

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -219,6 +261,67 @@ - (UIImage *)scaleImage:(UIImage *)image by:(id)scaleParam {
+      // Scaled image has close to same aspect ratio, updating image scale instead of resizing
+      // image.
+      return [self scaleImage:image
+                    withScale:(image.scale * (size.width / (image.size.width * image.scale)))];
```
will there be divide by 0 crash?

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -219,6 +261,67 @@ - (UIImage *)scaleImage:(UIImage *)image by:(id)scaleParam {
+- (UIImage *)scaleImage:(UIImage *)image toSize:(CGSize)size {
+  if (fabs(((int)image.size.width * image.scale) - size.width) > 0 ||
+      fabs(((int)image.size.height * image.scale) - size.height) > 0) {
+    if (fabs(image.size.width / image.size.height - size.width / size.height) < 1e-2) {
```
can you move this magic number as constant?

jokerttu replied:
removed the epsilon and replaced the functionality with a bit clearer code

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/ExtractIconFromDataTests.m:
```diff
@@ -0,0 +1,175 @@
+
+  UIImage *resultImage = [instance extractIconFromData:iconData registrar:mockRegistrar];
+  XCTAssertNotNil(resultImage);
+  [mockImageClass stopMocking];
```
im curious - when `mockImageClass` is deallocated, does it auto call the `stopMocking`?

jokerttu replied:
The mock should automatically calls stopMocking during its own deallocation ([ref](https://ocmock.org/reference/#:~:text=The%20mock%20automatically%20calls%20stopMocking%20during%20its%20own%20deallocation.)).
I removed these `[mockImageClass stopMocking]` lines.

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/ExtractIconFromDataTests.m:
```diff
@@ -0,0 +1,175 @@
+  XCTAssertNotNil(resultImage);
+}
+
+- (void)testScaleSizeToInt {
```
this seems to be a private function that shouldn't be exposed.

jokerttu replied:
These are exposed only for tests:

```
@interface FLTGoogleMapMarkerController (Test)
- (UIImage *)extractIconFromData:(NSArray *)iconData
                       registrar:(NSObject<FlutterPluginRegistrar> *)registrar;
- (CGSize)scaleSizeAndFloorToInt:(CGSize)size withFactor:(CGFloat)scale;
- (BOOL)isScalableWithScaleFactorFromSize:(CGSize)originalSize toSize:(CGSize)targetSize;
@end
```

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/ExtractIconFromDataTests.m:
```diff
@@ -0,0 +1,175 @@
+- (NSData *)convertImageToPNGData:(UIImage *)image;
+@end
+
+@implementation ExtractIconFromDataTests
```
is there a test to verify the resizing vs scaling threshold logic?

jokerttu replied:
I changed from calculating the difference between aspect ratios using epsilon to check the original size against target size with one pixel tolerance, to have better results.
Also added test cases to test this check.

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -160,52 +160,95 @@ - (UIImage *)extractIconFromData:(NSArray *)iconData
+    NSAssert((iconData.count == 3), @"'fromAssetImage' should have exactly 3 arguments. Got: %lu",
+             (unsigned long)iconData.count);
+    image = [UIImage imageNamed:[registrar lookupKeyForAsset:iconData[1]]];
+    id scaleParam = iconData[2];
```
can you type check here before passing into the helper function?

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -160,52 +160,95 @@ - (UIImage *)extractIconFromData:(NSArray *)iconData
+             (unsigned long)iconData.count);
+    @try {
+      FlutterStandardTypedData *byteData = iconData[1];
+      CGFloat screenScale = [[UIScreen mainScreen] scale];
```
i could be wrong - is `mainScreen` deprecated?

jokerttu replied:
@hellohuanlin how to proceed with this?

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios12/ios/RunnerTests/ExtractIconFromDataTests.m:
```diff
@@ -0,0 +1,345 @@
+  NSArray *testCases = @[
+    @{
+      @"originalSize" : [NSValue valueWithCGSize:CGSizeMake(100.0, 100.0)],
+      @"targetSize" : [NSValue valueWithCGSize:CGSizeMake(10.0, 100.0)],
```
can use shorthands like `@(CGSizeMake(100, 100))`

hellohuanlin reviewed (approved):
stamping for ios part. thanks for adding those tests.

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/README.md:
```diff
@@ -72,3 +74,27 @@ Indoor and building layers are still not available on the web. Traffic is.
+
+### Custom marker icons
+
+Images are not automatically scaled according to the `imageAspectRatio` value of the `BitmapDescriptor` in web, so the `size` parameter must be used to scale the marker when using the `AssetMapBitmap` or `BytesMapBitmap` methods.
```
The biggest issue (to this day) of custom markers on the web is that the maps SDK expects an image rendered @ "physical resolution", but then it needs help scaling it down to "logical resolution", so they look sharp (but not giant).

Is there any way of expressing that with the new primitives? Can users create a `BytesMapBitmap` image at 2x or 3x, then pass it with a size to the web? Do we still need this? Is there a good example for the web?

Is this behavior consistent with other platforms, or will people need some special code to render images that look sharp across all platforms?

jokerttu replied:
Scaling has some challenges on the web, and this PR currently supports scaling only through the size parameter, which is specified in logical pixels.

With AdvancedMarkers, it should be possible to specify an image element that has width parameter as a percentage, allowing for the scaling of the image based on the given pixel ratio. However, the Google Maps binding library does not yet include bindings for advanced markers: [Google Maps package on pub.dev](https://pub.dev/packages/google_maps) and [Advanced Markers documentation](https://developers.google.com/maps/documentation/javascript/advanced-markers/overview).

Scaling markers using percentages—which could be calculated from the pixel ratio and dpi—is not currently possible because of following problems:

Marker size is specified using two parameters:

```dart
  icon.size = <gmaps.Size object>;
  icon.scaledSize = <gmaps.Size object>;
```
where icon.size is applied to the div wrapper, and icon.scaledSize is applied to the image.

```dart
  external Size(
    num? width,
    num? height, [
    String? widthUnit, // ignore: unused_element
    String? heightUnit, // ignore: unused_element
  ]);
```
widthUnits can be used to give type of the unit (like `px`, `%`, `vh`, `vw`)

By default, Google Maps uses the image size for the div size, which is overridden if icon.size or icon.scaledSize is specified. It's not possible to set scaledSize without also overriding size. Giving size without knowing the right size of the image is not possible.

Earlier, I had a version where the image size was read from the image header data after loading, and it was automatically scaled using image scale and given pixel ratio. However, this method is not ideal by default, as it is not a very performance-efficient option, as the size need to be read from the image header. Note that with advanced markers it should be possible to to add onLoad event handler for image object given for the map, and tune the image element settings after the image is loaded.

So, with current version of google maps binding library, I don't see other options than support the size parameter on web. Of course in this case developers need to know the size of the asset; and for web assets this can be tricky without affecting the performance.

ditman replied:
> So if people know the image size, they can give it in proper size.
for example if the website has pixel ratio is 2.0 users just need to give 100x100 (physical pixel) image and add it as marker with size 50x50 (logical pixels).

As long as people can create their own marker as 100x100 (physical) and then tell google maps that it needs to be 50x50 (logical), I'm totally fine. I didn't see the separation of the `size` parameter on the new image classes, but they probably require a different process to draw the pixels, which then are passed to the map?

jokerttu replied:
Actually,
The previous method of utilizing the `image` package was not good because it used Dart code to access the image's dimensions (width and height) as discussed [here](https://github.com/flutter/plugins/pull/6805#discussion_r1095855056). Meaning the image was loaded and decoding the image in both Dart and the browser, which wasn't inefficient. 

Alternatively, by leveraging the [HTMLImageElement](https://pub.dev/documentation/web/latest/web/HTMLImageElement-extension-type.html) to load the image and extracting its dimensions on its onLoad callback function, performance should not be better. I'll test and betchmark if this approach works well and do not affect the overall performance of marker modification with the plugin. This approach must take care of the addition/modification of multiple markers to the map in a batch, avoiding the need to wait for each image to load sequentially.

jokerttu commented:
I'm currently focusing on updating test cases for the web implementation, following the implementation of automatic scaling. Related to these changes discussed [here](https://github.com/flutter/packages/pull/4055#discussion_r1521599738) using [HTMLImageElement](https://pub.dev/documentation/web/latest/web/HTMLImageElement-extension-type.html) to load image size on web.

I am also updating the constructors for `BytesMapBitmap` and `AssetMapBitmap`:

- For `AssetMapBitmap`, it will no longer require ImageConfiguration since this is primarily needed for mipmapping instructions. In scenarios where developers use assets without specifying an aspect ratio, it's generally preferable to default to an `imagePixelRatio` of 1.0. And developers need to instructed to give properly sized images for different 

- Similarly, `BytesMapBitmap` will default to an `imagePixelRatio` of 1.0, moving away from using the device's pixel density as the default setting. Developers aiming for higher resolution images should explicitly set the `imagePixelRatio` parameter or `size` parameter based on the desired aspect ratio. Defaulting images pixel ratio to devices pixel density, would change the marker size depending on the device, and would not have consistent results.

These updates are intended to enhance the usability and flexibility of these newly introduced classes.

I've converted this pull request to a `draft`, as I won't be able to complete this work today.

jokerttu commented:
@ditman automatic marker icon scaling without given size is now implemented for web as well using the onLoad event of [HTMLImageElement](https://pub.dev/documentation/web/latest/web/HTMLImageElement-extension-type.html). This implementation is visible in commit: https://github.com/flutter/packages/pull/4055/commits/3c2388832b519cb22d79c64d90374a571fb40250

@stuartmorgan 
Updated the documentation and constructors for `AssetMapBitmap` and `BytesMapBitmap` in commit: https://github.com/flutter/packages/pull/4055/commits/72f889eb23fad5ebcfc8aed77573279cd65dee15

Please review the documentation changes for these classes and methods as well.

As this PR also might cause upscaling on low-res images to keep the logical size same between the devices, it is important to have good documentation to have information how to control the sizing, with BitmapScaling.noScaling or with the imagePixelRatios or size parameters, which all give now really good and consistent control of the marker sizes between platforms.

ditman commented:
> it is important to have good documentation to have information how to control the sizing, with BitmapScaling.noScaling or with the imagePixelRatios or size parameters,

Flutter is able to pick the correct "Asset" at the right dpi, if it exists, from the assets directory ([Resolution-aware image assets](https://docs.flutter.dev/ui/assets/assets-and-images#resolution-aware)).

If possible, `AssetMapBitmap` should build off of that, right? (Then `BytesMapBitmap` would need "extra help" by specifying the `dpi` of each asset?)

jokerttu commented:
> > it is important to have good documentation to have information how to control the sizing, with BitmapScaling.noScaling or with the imagePixelRatios or size parameters,
> 
> Flutter is able to pick the correct "Asset" at the right dpi, if it exists, from the defined assets ([Resolution-aware image assets](https://docs.flutter.dev/ui/assets/assets-and-images#resolution-aware)).
> 
> If possible, `AssetMapBitmap` should build off of that, right? (Then `BytesMapBitmap` would need "extra help" by specifying the `dpi` of each asset?)

> > it is important to have good documentation to have information how to control the sizing, with BitmapScaling.noScaling or with the imagePixelRatios or size parameters,
> 
> Flutter is able to pick the correct "Asset" at the right dpi, if it exists, from the defined assets ([Resolution-aware image assets](https://docs.flutter.dev/ui/assets/assets-and-images#resolution-aware)).
> 
> If possible, `AssetMapBitmap` should build off of that, right? (Then `BytesMapBitmap` would need "extra help" by specifying the `dpi` of each asset?)

@ditman 
yes, but the current version also upscales assets if not given with proper dpi, and I think this need to be revisited and changed.

This PR already fixes a lot of issues, for example adds way to give logical size for assets that allows to render them wirh same exact logical size on each device and platform. 

Assets should always been drawn pixel-by-pixel by default, so current behaviour needs to be changed.
Current approach uses image pixelratio on native to draw the 30x30 image with pixelratio 1.0 as big as 90x90 with pixelratio 3.0.
Wirh this (currrent) approach the assets are by default drawn at same logical size on each device but this also causes upscaling, and visible pixels.

Aiming to render each asset automatically same logical size on each device is not the right default, and it should be done only with the size parameter, and automatic upscaling should be removed, as developers should be able to by default render the images on their natural resolution. Not sure if there is place for this upscaling at all. This of course means that same asset bitmap will be smaller on higher dpi displays, but developers should use mipmaps to mitigate this issue (or size parameter). Link to the doc for resolution aware assets should be added to the dartdocs? Might be that imagePixelRatio is not even needed on native side on any platform (I'll check if this can be dropped out)

Current approach follows how UIImage scale parameter works on iOS, where scale tells the images pixel ratio, causing it to be rendered with different size on the screen depending of the given value. The implementation should not change this value, and it should be kept as default (devices pixel ratio) for each UIImage.

For BytesMapBitmap the case not so simple. These assets could be for example images loaded from net. Developers should have way to scale these easily as they might not have access to different bitmaps for different dpi's. ImagePixelRatio might not be proper parameter for this. Instead this should have scale parameter which can be used to upscale (or downscale) these bitmaps. This allows them to use the devices dpi as scale to keep the asset upscaled to same size on each device. Developers can use size parameter as well.

So, my suggestion is that I do following changes:
- I remove automatic upscaling to dpi by default (on iOS UIImage scale should be kept as default)
- imagePixelRatio parameter should be replaced with scale parameter for both AssetMapBitmap and BytesMapBitmap allowing developers to easily upscale (or downscale) bitmaps if they want without knowing the images size. Scale is by default 1.0.
- BitmapScaling enum could be removed as there is now automatic scaling.
- Tests and documentation updates

I think the updated behaviour should be better and simpler, eventhough this change would mean that the asset size would differ wirh different dpi's but as said, sharp rendering is more important, and developer always can use size parameter (or scale) if they want.

On web the image width and height need to be always devided with devices pixel ratio as it render everything too big by default on high dpi displays; this is easy to fix now when we have way to detect image size on the web plstform.

Sorry for the tuning. This change should not take long to fix as this is quite small change and won't cause big changes on native code either.

ditman commented:
> On web the image width and height need to be always devided with devices pixel ratio as it render everything too big by default

More precisely, the web uses the "logical size" of an asset (physical size / dpr) for everything, but the images need to be constructed at the "physical size".

For a web developer, the weirdness is having to create their asset at 2x when DPI is > 1 (see [this recent issue](https://github.com/flutter/flutter/issues/143398), for example!)

If you update the docs more, please describe how web devs can have sharp images (I think the web is better at downscaling than upscaling, but even then, having the right size for the current screen DPI is most important)

((Best image size for the current DPI == what iOS calls mip-mapping?))

jokerttu commented:
> > On web the image width and height need to be always devided with devices pixel ratio as it render everything too big by default
> 
> More precisely, the web uses the "logical size" of an asset (physical size / dpr) for everything, but the images need to be constructed at the "physical size".

Yes, exactly, 
I was wrong with this approach: https://github.com/flutter/packages/pull/4055#issuecomment-1998720327.
Currently implemented behaviour works same way as flutter is rendering assets with certain scales (scale is same as imagePixelRatio). If developers only give asset with imagePixelRatio/scale of 1.0 it is **rendered upscaled** on flutter if devices pixel ratio is >1, and therefore for flutter developers I think the current approach matches the behaviour.
For BitmapScalingMethod, current behaviour give two options, `auto` and `noScaling` where noScaling means that the images are drawn directly on platform without scaling, and this causes some issues. On iOS and android images are drawn pixel-by-pixel but as you said, on web the image is drawn differently, it always scales the images with the device pixel ratio by default (and this is understandable as all websites needed to work with first retina displays as same size as 1.0 pixel ratio displays). 
With current approach this means that if developers want to render the image `pixel-by-pixel`, the only way to do this on all platforms is to give the devices pixel ratio as a value for imagePixelRatio. 

I am trying to find best API for scaling, and might rename the `noScaling` as `platformDefault`,  as this describes better what is happening (like web really does scaling in this situation). This option is given as it is more performant option if performance need to be improved by certain use cases.

I'm also changing how logical `size` is given for the bitmaps (assets and bitmaps), as it currently does not preserve aspect ratio. Instead of `size`, developers should be able to give optional `width` and `height` parameters. If only width is given, images are resized to this and aspect ratio is preserved and wise versa. If both are given the image is rendered using the exact size given. Flutter `ResizeImage` class uses following policy https://api.flutter.dev/flutter/painting/ResizeImagePolicy.html, and I am going to follow the `exact` policy with marker image scaling. The fit policy will not be implemented, but it can be added quite easily with separate "MapBitMapResizePolicy" enum holding `exact` and `fit` policies if needed.

> 
> For a web developer, the weirdness is having to create their asset at 2x when DPI is > 1 (see [this recent issue](https://github.com/flutter/flutter/issues/143398), for example!)

Developers should always provide the assets/images for certain device dpi:s.
This PR already unifies how this can be controlled with `imagePixedRatio` and `size` (note: size will be replaced with `width` and `height`) on each platform, meaning that on https://github.com/flutter/flutter/issues/143398 the issue can be solved only by providing proper imagePixelRatio for the asset and it should be rendered automatically with proper `logical size`. For web, it should be recommended to always give assets and bitmaps at 2.0 scale/imagePixelRatio, as then they will be always sharp on all device pixel ratios.
> 
> If you update the docs more, please describe how web devs can have sharp images (I think the web is better at downscaling than upscaling, but even then, having the right size for the current screen DPI is most important)

The most important thing is to improve the docs, and maybe even highlight separately how web markers images should be provided. Developers could also provide assets also (or only) in `2.0x` folder to ensure sharp rendering on each platform.
For image generation the device pixel ratio should be used to render properly scaled image.
> 
> ((Best image size for the current DPI == what iOS calls mip-mapping?))
> 
> ((PS: thanks for taking care of this, it's a big source of issues for web users!))

PS. when flutter is finding proper asset size for device pixel ratios higher than 1 but less than 2 (>1, <=2), it will always select 2, as upscaling from 1 to for example 1.4 does not look good. On higher scales the closest is selected.

jokerttu commented:
Sorry for a long period from last update on this.
Refactored the behaviour to match the review comments given in history.

One of the main concerns was the sizing without preserving image aspect ratio, this is now fixed as instead of `size`, users can now init AssetMapBitmap and BytesMapBitmap with optional `width` and `height` parameters. If only one of these are given the image is scaled preserving its aspect ratio on each platform. Also imagePixelRatio is supported for both types (used if width and height is not given).

Serialization for `asset` and `bytes` BitmapDescription types are updated for cleared parser on native Android and iOS implementations.

Updated example apps to allow testing the assets with only width or height or both.

Added helper wrapper methods to the BitmapDescriptor class.

Hopefully this is now better and clearer API for the developers.

hellohuanlin reviewed (approved):
I think I previously stamped the iOS part, but somehow the stamp is gone. Re-stamp here. 

Also could you address my previous comments that's not addressed yet?

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:
```diff
@@ -262,20 +271,178 @@ gmaps.Size? _gmSizeFromIconConfig(List<Object?> iconConfig, int sizeIndex) {
+            height: bitmapDescriptor.height,
+            imagePixelRatio: bitmapDescriptor.imagePixelRatio,
+            url: url,
+            cacheKey: bytes.hashCode.toString(),
```
The URL of the Blob should be "unique enough" for this cacheKey. I think cacheKey can be removed.

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:
```diff
@@ -262,20 +271,178 @@ gmaps.Size? _gmSizeFromIconConfig(List<Object?> iconConfig, int sizeIndex) {
+        break;
+    }
+    return icon;
+  }
```
This method is too large, I think it could be split into two separate functions.

Also the top half can be simplified greatly:

1. Compute the URL depending on the type of `bitmapDescriptor`.
2. Create the Icon (shared code)
3. if (bitmapDescriptor.bitmapScaling == MapBitmapScaling.auto) call _setIconSize.

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:
```diff
@@ -262,20 +271,178 @@ gmaps.Size? _gmSizeFromIconConfig(List<Object?> iconConfig, int sizeIndex) {
+  } else {
+    assert(
+        url.isNotEmpty, 'URL must not be empty when calculating dimensions.');
+    final Size? bitmapSize = await _getBitmapSize(url, cacheKey);
+
+    // If bitmap size is null, the image is invalid,
+    // and the icon size is not set.
+    if (bitmapSize != null) {
+      double targetWidth = bitmapSize.width;
+      double targetHeight = bitmapSize.height;
+
+      if (width == null && height == null) {
+        // Adjust the target size based on the image pixel ratio.
+        targetWidth /= imagePixelRatio;
+        targetHeight /= imagePixelRatio;
+      } else {
+        final double aspectRatio = bitmapSize.width / bitmapSize.height;
+        targetWidth = width ?? (height ?? bitmapSize.height) * aspectRatio;
+        targetHeight = height ?? (width ?? bitmapSize.width) / aspectRatio;
+      }
```
This `_setIconSize` is doing way too many things IMO. Please separate the `_getBitmapSize` call from here, so it's done before calling `_setIconSize`. `_setIconSize` should be super dumb, just receive width, height and icon.

jokerttu replied:
Refactored

ditman reviewed (changes requested):
[web stuff] I like the new code paths for the more advanced marker bitmap types, but please, split those in separate methods rather than making the current methods handle all 4 possible cases!

The problems are in convert.dart: `_setIconSize` is also getting the size and doing a bunch of checks on how to use it, and `_gmIconFromBitmapDescriptor` is now close to 100 LOC.

I think those can be split at the place they're called by checking the type of the Bitmap, and if it's modern -> go do the modern thing, but if it's "json-backed" -> do the old stuff.

The lines of code would be about the same, but it'd be cleaner. It'd also make it easier to remove the "old" code paths when the deprecated types disappear.

PTAL!

jokerttu reviewed (commented):
Updated the structure as requested with few comments.
Added abstract MapBitmap super class for AssetMapBitmap and BytesMapBitmap to allow shared functionality for Web implementation that uses these classes directly.

I was not able to figure out the deprecated mainscreen issue totally, this is discussed here https://github.com/flutter/packages/pull/4055/files#r1518139100.

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:
```diff
@@ -262,20 +271,178 @@ gmaps.Size? _gmSizeFromIconConfig(List<Object?> iconConfig, int sizeIndex) {
+  // Wait for the onLoad or onError event.
+  await Future.any(<Future<Event>>[image.onLoad.first, image.onError.first]);
+
+  if (image.width == 0 || image.height == 0) {
+    // Complete with null for invalid images.
+    completer.complete(null);
+  } else {
+    // Complete with the image size for valid images.
+    completer.complete(Size(image.width.toDouble(), image.height.toDouble()));
+  }
```
This method should be equivalent to yours, I think:

```dart
// _bitmapSizeCache should be a Map<String, Size?>

Future<Size?> _fetchBitmapSize(String url) async {
  if (!_bitmapSizeCache.containsKey(url)) {
    final HTMLImageElement image = HTMLImageElement()..src = url;

    // Wait for the onLoad or onError event.
    await Future.any(<Future<Event>>[image.onLoad.first, image.onError.first]);

    if (image.width == 0 || image.height == 0) {
      // Complete with null for invalid images.
      _bitmapSizeCache[url] = null;
    } else {
      // Complete with the image size for valid images.
      _bitmapSizeCache[url] = Size(image.width.toDouble(), image.height.toDouble());
    }
  }
  return _bitmapSizeCache[url]; // May be null
}
```

By making the method `async` and using an `await` in the middle of it, you already have a "free completer" that will be fulfilled when you `return`.

jokerttu replied:
This is not a same. 
As _fetchBitmapSize can be called multiple times before 
`await Future.any(<Future<Event>>[image.onLoad.first, image.onError.first])` is resolved and the resolved size is cached with url.
This is because multiple markers are handled in paraller for speed.

jokerttu replied:
This would work:

```dart
final Map<String, Future<Size?>> _bitmapSizeFutureCache =
    <String, Future<Size?>>{};

...

/// Fetches the size of the bitmap from a given URL and caches the result.
///
/// This method attempts to fetch the image size for a given [url].
Future<Size?> _fetchBitmapSize(String url) async {
  final HTMLImageElement image = HTMLImageElement()..src = url;

  // Wait for the onLoad or onError event.
  await Future.any(<Future<Event>>[image.onLoad.first, image.onError.first]);

  if (image.width == 0 || image.height == 0) {
    // Complete with null for invalid images.
    return null;
  }

  // Complete with the image size for valid images.
  return Size(image.width.toDouble(), image.height.toDouble());
}
```

And use this like this

```dart
    final Size? bitmapSize = await _bitmapSizeFutureCache.putIfAbsent(url, () {
      return _fetchBitmapSize(url);
    });
```

jokerttu replied:
Committed, as same future can be read multiple times, this works as well as completer, and is much simpler; thanks!

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:
```diff
@@ -262,20 +271,178 @@ gmaps.Size? _gmSizeFromIconConfig(List<Object?> iconConfig, int sizeIndex) {
+///
+/// - Parameters:
+/// [url] is the URL of the image for which the size is being fetched.
+/// [cacheKey] is a unique key for each image used for caching the size calculation result.
```
Yeah, this should be fine, especially because the cache is quite short-lived.

jokerttu commented:
@hellohuanlin 
> Could you address my previous comments that's not addressed yet?

Sorry, missed some earlier; should be now fixed but this one is still open for discussion: https://github.com/flutter/packages/pull/4055#discussion_r1568646976

Also notice that the logic for resizing is changed from optional size to optional width and height values allowing easier way to preserve aspect ratio.

ditman commented:
I don't have more comments on the web implementation, thanks @jokerttu!!

ditman reviewed (approved):
[web] Web changes look good to me; thanks @jokerttu!

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -379,6 +587,10 @@ - (void)isInfoWindowShownForMarkerWithIdentifier:(NSString *)identifier
 }
 
+- (CGFloat)getScreenScale {
+  return self.mapView.traitCollection.displayScale;
```
@hellohuanlin is this acceptable fix for getting screenScale for now. See comment here: https://github.com/flutter/packages/pull/4055#discussion_r1568646976
This is currently only open review issue before splitting this tasks to sub PR's

jokerttu replied:
@stuartmorgan this has been waiting to be resolved last few weeks?
I would like to start plitting this to sub-pr's but this one issue is still blocking the process. Any way to have somekind of solution for this?
This is a bit bigger issue as the markers are initialized before the virw is attached to the parent view. The whole maps initialialization should be changed a bit. Or platformview factory should be able to get some info about the screen where it will be attached to, but that is not cyrrently possible (if the screen scale is not sent in platform view creation params from the dart implementation).

stuartmorgan replied:
I would file an issue explaining the issue, add a comment on this line with a TODO pointing to the issue, and proceed with that. As you pointed out, the cases where this won't work correctly should be uncommon, and likely require Flutter SDK-level changes to correctly handle.

reidbaker commented:
@jokerttu is this ready for a another pass from me?

jokerttu commented:
As platform interface changes are already landed (https://github.com/flutter/packages/pull/6687), deprecating old methods, I have went forward on creating PR for platform implementations of this federated plugin here: https://github.com/flutter/packages/pull/6826.

@reidbaker could you continue the process on this PR.

jokerttu commented:
Converted as draft to avoid unnecessary reviews on this PR before platform implementations are landed.
See the platform implementation portion of this PR here: https://github.com/flutter/packages/pull/6826

This PR will be updated and marked ready for review with only app-facing plugin changes after the platform impls are landed.

jokerttu commented:
This PR now contains only the app-facing api changes in google_maps_flutter package.

@stuartmorgan what would be correct process to deprecate the old methods under BitmapDescriptor class.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter/CHANGELOG.md:
```diff
@@ -9,7 +10,6 @@
 * Updates README and example app to reflect that the minimum supported version
   is now iOS 14. Removal of support for iOS 12 and 13 was necessary due to the
   upcoming App Store privary manifest requirements.
-
```
Nit: This shouldn't be removed.

stuartmorgan reviewed (approved):
LGTM!

You can send me a PR to re-deprecate the other bitmap constructors, and we'll land that after landing this one. (It should be safe to land before this too, but I think it's less confusing if have already re-exported the new types.)
